### ISSUE TITLE
IDE0251 error in error list has no title

### ISSUE THREAD

**Issue by Youssef1313** [dotnet/roslyn#67949]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/IDE0251-error-in-error-list-has-no-descr/10312916)._

[severity:It bothers me. A fix would be nice] I just opened my project and noticed in the Error List window I have two IDE0251 errors with no descriptions. If I click to go to the location of the error, I see code is fine. Also, the link to the documentation around that error seems to be broken, I get a 404 on it (https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0251)

#### Feedback Bot on 3/16/2023, 05:46 PM:
(private comment, text removed)

#### Feedback Bot on 3/17/2023, 06:31 AM:
(private comment, text removed)

### Original Solutions
(no solutions)

**Comment by Youssef1313** [comment]
The original ticket was incorrectly moved to roslyn-analyzers repo https://github.com/dotnet/roslyn-analyzers/issues/6552.

**Comment by DanTravison** [comment]
I saw a number of these with no description in my build and all were reporting for struct Properties with public get accessors .  In all cases, the suggested fix is to make the member read-only. 

I tried refactoring to make the structure read-only which eliminated the warnings but implementation and usage becomes rather convoluted since the methods that normally updated the current instance in place now need to create a copy with the updated values.  

Is this an analyzer issue or am I missing something?

The project is configured with the following:
<EnableNETAnalyzers>true</EnableNETAnalyzers>
<AnalysisLevel>latest</AnalysisLevel>

I'm building with Microsoft Visual Studio Community - Version 17.6.0 Preview 4.0

I don't see these when I build with Microsoft Visual Studio Community - Version 17.5.4

Here are a few examples:

public float X
{
    get => _x; <--
}

get  <--
{
    return new PointF(_x, _y);
}


public Color ToColor()  <--
{
    double chroma = Brightness * Saturation;
    double hue = Hue / 60;
    double x = chroma * (1 - Math.Abs((hue % 2) - 1));
    double m = Brightness - chroma;

    return HSLColor.ToColor(chroma, hue, x, m, Alpha);
}

**Comment by CyrusNajmabadi** [comment]
@DanTravison can you update your code sample with ```` ``` ```` markdown flags?

**Comment by CyrusNajmabadi** [comment]
> I saw a number of these with no description in my build and all were reporting for struct Properties with public get accessors . In all cases, the suggested fix is to make the member read-only.

Yes.  Making the property readonly will help when calling it on a readonly instance of your struct.  The compiler will now no longer have to copy the instance in that case as it can be sure that calling the property directly is safe (as it can't mutate the instance).

**Comment by DanTravison** [comment]
I have few issues with the warning:
1: The properties only have a get accessory. Isn't that, by definition, a readonly property? 
2: The underlying fields are explicitly declared as private. 
I don't see the reasoning behind adding readonly declaration to a read-only property. 
3: In one case, it is flagging a method that returns a value based on a calculation of a private field. I can't declare that read-only and the field cannot be declared readonly because it is used as an accumulator.

**Comment by CyrusNajmabadi** [comment]
> The properties only have a get accessory. Isn't that, by definition, a readonly property?

Nope.  Consider the following:

```c#
int Prop { get { this.x++; return this.x; }
```

> 2: The underlying fields are explicitly declared as private.

There's no cross symbol understanding.  At the user site it has to assume the worst.

> I don't see the reasoning behind adding readonly declaration to a read-only property.

To ensure the compiler doesn't copy your structs when you access properties off of them if the instance is in a readonly location.

> In one case, it is flagging a method that returns a value based on a calculation of a private field. I can't declare that read-only

Why not?  :)  Looks fine to be made a readonly prop or method.

**Comment by CyrusNajmabadi** [comment]
> All of the above begs the question of why the analyzer output does not include a description

That's a bug.

>  and why the link results in a 404 error. 

Also a bug.  This issue tracks fixing those issues.  I'm just trying to answer your personal questions about your scenario.  :)

**Comment by DanTravison** [comment]
> declaring a method/property as read-only

I must have missed that language change.  Thanks for pointing that out.

> There's no cross symbol understanding. At the user site it has to assume the worst.

This is what led to my confusion. When I see [IDE0044](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0044) reported for class fields, I don't hesitate to use the guidance since the page clearly states why and I've confirmed it enough to trust the recommendation.  With IDE0251, there is no explanation and I'm reluctant to make changes I don't understand.

With your explanation,  it's obvious there's a case I didn't consider.

Thanks.

**Comment by MaceWindu** [comment]
```cs
record struct SomeRecord(int X)
{
        // IDE0251: <empty>
	public bool Equals(CacheKey? other) => false;

        // IDE0251: <empty>
	public override int GetHashCode() => 0;
}
```

**Comment by PyrateAkananto** [comment]
I am affected by the same issue.

Unity project
Visual Studio Code 1.78.2
C# for Visual Studio Code (powered by OmniSharp) 1.25.7
Microsoft.Unity.Analyzers 1.15.0

Empty text for IDE0251 in "PROBLEMS" window, see attached screenshot, caused by the following code:

```
public struct DummyStruct
{
    public int DummyValue;

    // the "get" in the following line causes an IDE0251 notification
    public int GetDummyValue { get => DummyValue; set => DummyValue = value; }
}
```

![IDE0251-empty-text](https://github.com/dotnet/roslyn/assets/11444373/acd17311-4687-4073-8a35-886aa6f8492c)

**Comment by kkm000** [comment]
**REPRO:** I'm also getting the IDE0251 with no message, but no suggested fix after VS2022 upgrade to Version 17.6.2. This is the code fragment that's frowned upon (^^ mark the dotted underscores):
```c#
record struct VoiceMarker(uint Start, uint Length) : IComparable<VoiceMarker> {
  public uint End => Start + Length;
//                ^^
  public int CompareTo(VoiceMarker v) =>
//           ^^
    Comparer<uint>.Default.Compare(Start, v.Start);  // Length is irrelevant.
}
```
The project file is unremarkable:
```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net7.0</TargetFramework>
    <ImplicitUsings>disable</ImplicitUsings>
    <Nullable>enable</Nullable>
    <AllowUnsafeBlocks>True</AllowUnsafeBlocks>
    <ProduceReferenceAssembly>False</ProduceReferenceAssembly>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="DotNext.IO" Version="4.12.0" />
    <PackageReference Include="DotNext.Unsafe" Version="4.12.0" />
    <PackageReference Include="System.CommandLine" Version="2.0.0-beta4.22272.1" />
  </ItemGroup>
</Project>
```

**Comment by kkm000** [comment]
To suppress the message, add to `.editorconfig`:

```ini
[*.cs]
# https://github.com/dotnet/roslyn/issues/67949
dotnet_diagnostic.IDE0251.severity = none
```

**Comment by Jens-G** [comment]
> I am affected by the same issue.
> Empty text for IDE0251 in "PROBLEMS" window, see attached screenshot, caused by the following code:

Plus is emits the "make it readonly" for a property that **has a setter**. Making that one actually `readonly` surely makes IDE0251 go away, but gives "CS1604 cannot set write protected field" instead. 

To me it looks pretty much as if the emitted IDE0251 is also incorrect in some cases ...

**Comment by grimreaper38uk** [comment]
Has this still not been fixed??
VS 2022 v17.6.5

**Comment by cremor** [comment]
I think this was fixed by #67664, so it should be fixed in VS 17.7
The missing documentation page is tracked in dotnet/docs#30361

**Comment by mavasani** [comment]
Closing this out as the original issue reported here has been fixed with #67664. If there are any additional functional issues with this rule, kindly file separate issues. Thanks!
